gopls: packages.load error: JSON decoding failed : invalid character 'x' in string escape code

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I input command `brew upgrade zed` and zed version from 0.132.2 to 0.134.4.
this is error "gopls: packages.load error: JSON decoding failed : invalid character 'x' in string escape code " when I open exist go project in zed.


### Environment

Zed: v0.134.4 (Zed)
OS: macOS 13.0.1
Memory: 16 GiB
Architecture: x86_64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

<img width="1412" alt="image" src="https://github.com/zed-industries/zed/assets/13704743/3e2e219e-4968-4f9d-979f-3f53c9bc2287">


### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

2024-05-10T14:34:48+08:00 [ERROR] crates/gpui/src/elements/img.rs:344: Client(Error { kind: Timeout, context: None, source: Some(Error { description: "Timeout was reached", code: 28, extra: None }), source_type: Some("curl::error::Error"), local_addr: None, remote_addr: None })
2024-05-10T14:34:48+08:00 [ERROR] crates/gpui/src/elements/img.rs:344: Client(Error { kind: Timeout, context: None, source: Some(Error { description: "Timeout was reached", code: 28, extra: None }), source_type: Some("curl::error::Error"), local_addr: None, remote_addr: None })
2024-05-10T14:35:06+08:00 [ERROR] crates/gpui/src/elements/img.rs:344: Client(Error { kind: Timeout, context: None, source: Some(Error { description: "Timeout was reached", code: 28, extra: None }), source_type: Some("curl::error::Error"), local_addr: None, remote_addr: None })
2024-05-10T14:35:07+08:00 [WARN] Generic lsp request to gopls failed: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/suggest.go
2024-05-10T14:35:07+08:00 [ERROR] crates/project/src/project.rs:7009: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/suggest.go
2024-05-10T14:35:07+08:00 [ERROR] crates/gpui/src/elements/img.rs:344: Client(Error { kind: Timeout, context: None, source: Some(Error { description: "Timeout was reached", code: 28, extra: None }), source_type: Some("curl::error::Error"), local_addr: None, remote_addr: None })
2024-05-10T14:35:07+08:00 [ERROR] crates/gpui/src/elements/img.rs:344: Client(Error { kind: Timeout, context: None, source: Some(Error { description: "Timeout was reached", code: 28, extra: None }), source_type: Some("curl::error::Error"), local_addr: None, remote_addr: None })

log file
`2024-05-30T14:18:39+08:00 [ERROR] unexpected item event after pane was dropped
2024-05-30T14:19:08+08:00 [INFO] Initializing default prettier with plugins {}
2024-05-30T14:19:08+08:00 [INFO] starting language servers for Lua: lua-language-server
2024-05-30T16:40:01+08:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-05-30T16:40:01+08:00 [INFO] Initializing default prettier with plugins {}
2024-05-30T16:40:01+08:00 [INFO] starting language servers for Go: gopls
2024-05-30T16:40:01+08:00 [INFO] starting language server "gopls", path: "/Users/xsh/fm/fm-suggest", id: 10
2024-05-30T16:40:01+08:00 [ERROR] crates/workspace/src/persistence/model.rs:313: No worktree for path: "/Users/xsh/.config/zed/settings.json"
2024-05-30T16:40:02+08:00 [INFO] found user-installed language server for Go. path: "/Users/xsh/go/bin/gopls", arguments: ["-mode=stdio"]
2024-05-30T16:40:02+08:00 [INFO] starting language server. binary path: "/Users/xsh/go/bin/gopls", working directory: "/Users/xsh/fm/fm-suggest", args: ["-mode=stdio"]
2024-05-30T16:40:02+08:00 [INFO] Language server with id 10 sent unhandled notification window/logMessage:
{
  "type": 1,
  "message": "2024/05/30 16:40:02 creating temp dir: mkdir /var/folders/d9/h1ryx87j3pq8gcx7knm2xssh0000gn/T/\n\u001b[?2004l/gopls-6762.1: no such file or directory\n"
}
2024-05-30T16:40:02+08:00 [WARN] Generic lsp request to gopls failed: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-05-30T16:40:02+08:00 [ERROR] crates/project/src/project.rs:7172: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-05-30T16:40:03+08:00 [WARN] Generic lsp request to gopls failed: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-05-30T16:40:03+08:00 [ERROR] crates/project/src/project.rs:7172: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-05-30T16:40:07+08:00 [INFO] Initializing default prettier with plugins {}
2024-05-30T16:40:07+08:00 [INFO] starting language servers for Go Mod: gopls
2024-05-30T16:42:15+08:00 [ERROR] crates/lsp/src/lsp.rs:779: oneshot canceled
2024-05-30T16:42:15+08:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-05-30T16:42:15+08:00 [INFO] Initializing default prettier with plugins {}
2024-05-30T16:42:15+08:00 [INFO] starting language servers for Go: gopls
2024-05-30T16:42:15+08:00 [INFO] starting language server "gopls", path: "/Users/xsh/fm/fm-suggest", id: 11
2024-05-30T16:42:15+08:00 [ERROR] crates/workspace/src/persistence/model.rs:313: Deserializer does not exist for item kind: diagnostics
2024-05-30T16:42:16+08:00 [INFO] found user-installed language server for Go. path: "/Users/xsh/go/bin/gopls", arguments: ["-mode=stdio"]
2024-05-30T16:42:16+08:00 [INFO] starting language server. binary path: "/Users/xsh/go/bin/gopls", working directory: "/Users/xsh/fm/fm-suggest", args: ["-mode=stdio"]
2024-05-30T16:42:16+08:00 [INFO] Language server with id 11 sent unhandled notification window/logMessage:
{
  "type": 1,
  "message": "2024/05/30 16:42:16 creating temp dir: mkdir /var/folders/d9/h1ryx87j3pq8gcx7knm2xssh0000gn/T/\n\u001b[?2004l/gopls-6871.1: no such file or directory\n"
}
2024-05-30T16:42:16+08:00 [WARN] Generic lsp request to gopls failed: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-05-30T16:42:16+08:00 [ERROR] crates/project/src/project.rs:7172: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-05-30T16:42:21+08:00 [ERROR] crates/lsp/src/lsp.rs:779: oneshot canceled`

I update zed 0.137.6 today, the problem no fixed, log file is
```
2024-06-03T10:09:18+08:00 [INFO] ========== starting zed ==========
2024-06-03T10:09:18+08:00 [INFO] Opening main db
2024-06-03T10:09:18+08:00 [INFO] Using git binary path: Some("/Applications/Zed.app/Contents/MacOS/git")
2024-06-03T10:09:18+08:00 [INFO] set environment variables from shell:/usr/local/bin/fish, path:/opt/go/bin:/Users/xsh/go/bin:/opt/homebrew/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin
2024-06-03T10:09:18+08:00 [ERROR] crates/theme/src/settings.rs:432: theme not found: Catppuccin Macchiato
2024-06-03T10:09:18+08:00 [INFO] extensions updated. loading 5, reloading 0, unloading 0
2024-06-03T10:09:18+08:00 [INFO] Opening main db
2024-06-03T10:09:18+08:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-06-03T10:09:19+08:00 [INFO] Opening main db
2024-06-03T10:09:19+08:00 [INFO] set status on client 0: Authenticating
2024-06-03T10:09:19+08:00 [INFO] Opening main db
2024-06-03T10:09:19+08:00 [INFO] set status on client 79320: Connecting
2024-06-03T10:09:19+08:00 [INFO] Initializing default prettier with plugins {}
2024-06-03T10:09:19+08:00 [INFO] starting language servers for Go: gopls
2024-06-03T10:09:19+08:00 [INFO] starting language server "gopls", path: "/Users/xsh/fm/fm-suggest", id: 1
2024-06-03T10:09:19+08:00 [INFO] Initializing default prettier with plugins {}
2024-06-03T10:09:19+08:00 [INFO] starting language servers for Go: gopls
2024-06-03T10:09:19+08:00 [ERROR] crates/workspace/src/persistence/model.rs:313: Deserializer does not exist for item kind: diagnostics
2024-06-03T10:09:19+08:00 [INFO] found user-installed language server for Go. path: "/Users/xsh/go/bin/gopls", arguments: ["-mode=stdio"]
2024-06-03T10:09:19+08:00 [INFO] starting language server. binary path: "/Users/xsh/go/bin/gopls", working directory: "/Users/xsh/fm/fm-suggest", args: ["-mode=stdio"]
2024-06-03T10:09:19+08:00 [INFO] Language server with id 1 sent unhandled notification window/logMessage:
{
  "type": 1,
  "message": "2024/06/03 10:09:19 creating temp dir: mkdir /var/folders/d9/h1ryx87j3pq8gcx7knm2xssh0000gn/T/\n\u001b[?2004l/gopls-56436.1: no such file or directory\n"
}
2024-06-03T10:09:19+08:00 [INFO] Node runtime install_if_needed
2024-06-03T10:09:19+08:00 [INFO] Initializing default prettier with plugins {}
2024-06-03T10:09:19+08:00 [INFO] starting language servers for Go: gopls
2024-06-03T10:09:20+08:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2024-06-03T10:09:21+08:00 [INFO] starting language server. binary path: "/Users/xsh/Library/Application Support/Zed/node/node-v18.15.0-darwin-x64/bin/node", working directory: "/", args: ["/Users/xsh/Library/Application Support/Zed/copilot/copilot-v0.5.0/dist/agent.js", "--stdio"]
2024-06-03T10:09:21+08:00 [INFO] add connection to peer
2024-06-03T10:09:21+08:00 [INFO] add_connection;
2024-06-03T10:09:21+08:00 [INFO] waiting for server hello
2024-06-03T10:09:21+08:00 [INFO] got server hello
2024-06-03T10:09:21+08:00 [INFO] set status to connected (connection id: ConnectionId { owner_id: 0, id: 0 }, peer id: PeerId { owner_id: 440, id: 1820744 })
2024-06-03T10:09:21+08:00 [INFO] set status on client 79320: Connected { peer_id: PeerId { owner_id: 440, id: 1820744 }, connection_id: ConnectionId { owner_id: 0, id: 0 } }
2024-06-03T10:09:21+08:00 [ERROR] crates/lsp/src/lsp.rs:360: cannot read LSP message headers
2024-06-03T10:09:21+08:00 [ERROR] crates/lsp/src/lsp.rs:382: Broken pipe (os error 32)
2024-06-03T10:09:21+08:00 [ERROR] crates/lsp/src/lsp.rs:779: server shut down
2024-06-03T10:09:24+08:00 [WARN] Generic lsp request to gopls failed: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-06-03T10:09:24+08:00 [ERROR] crates/project/src/project.rs:7172: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-06-03T10:09:25+08:00 [WARN] Generic lsp request to gopls failed: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go
2024-06-03T10:09:25+08:00 [ERROR] crates/project/src/project.rs:7172: no package metadata for file file:///Users/xsh/fm/fm-suggest/internal/data/reply.go

```

Hey @notpeter 👋 
These are paths for the regular start (Application): 
```
/Users/.../go_path/bin 
/Users/valery/.cargo/bin 
/opt/homebrew/bin 
/usr/local/bin 
/System/Cryptexes/App/usr/bin 
/usr/bin 
/bin 
/usr/sbin 
/sbin 
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin /Library/Apple/usr/bin
```
And these are when started from the terminal:
```
/Users/.../go_path/bin 
/Users/valery/.cargo/bin 
/opt/homebrew/bin 
/usr/local/bin 
/System/Cryptexes/App/usr/bin 
/usr/bin 
/bin 
/usr/sbin 
/sbin 
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin /Library/Apple/usr/bin 
/Applications/kitty.app/Contents/MacOS
```

The repeated error in logs is: `[ERROR] no package metadata for file file` + filepath.
All Fish env variables are in the `.config/fish/fish_variables` file.

> I'm curious where you are adding ~/go/bin to your path for Fish:

I'm using only `fish_add_path`. No additional envs configuration in the `config.fish`

> Or: did you run fish_add_path $HOME/go/bin once in an interactive shell and are relying on something in ~/.config/fish/fish_variables that looks like.

Yep, correct.

@notpeter the problem persists for me with Zed 0.161.2

Status at the bottom keeps saying "Error loading workspace: packages.Load error JSON decoding failed ..."

And the log doesn't have much more than this:
```
2024-11-14T17:43:51.751139-05:00 [WARN] Generic lsp request to gopls failed: no package metadata for file <filename>
2024-11-14T17:43:52.901271-05:00 [WARN] Generic lsp request to gopls failed: no package metadata for file <filename>
2024-11-14T17:43:52.91782-05:00 [ERROR] no package metadata for file <filename>
```

Could you please reopen this issue as it's very much not fixed. Note that starting from command line workaround still works, and that the log contains the same warns / error as above.

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -246,91 +246,44 @@ async fn load_shell_environment(
     Option<EnvironmentErrorMessage>,
 ) {
     use crate::direnv::{DirenvError, load_direnv_environment};
-    use std::path::PathBuf;
-    use util::parse_env_output;
+    use util::shell_env;
 
     fn message<T>(with: &str) -> (Option<T>, Option<EnvironmentErrorMessage>) {
         let message = EnvironmentErrorMessage::from_str(with);
         (None, Some(message))
     }
 
-    const MARKER: &str = "ZED_SHELL_START";
     let Some(shell) = std::env::var("SHELL").log_err() else {
         return message("Failed to get login environment. SHELL environment variable is not set");
     };
-    let shell_path = PathBuf::from(&shell);
-    let shell_name = shell_path.file_name().and_then(|f| f.to_str());
 
-    // What we're doing here is to spawn a shell and then `cd` into
-    // the project directory to get the env in there as if the user
-    // `cd`'d into it. We do that because tools like direnv, asdf, ...
-    // hook into `cd` and only set up the env after that.
-    //
+    let dir_ = dir.to_owned();
+    let captured = match smol::unblock(move || shell_env::capture(shell, Some(dir_))).await {
+        Err(err) => {
+            util::log_err(&err);
+            return message(
+                "Failed to spawn login shell to source login environment variables. See logs for details",
+            );
+        }
+        Ok(Err(err)) => {
+            util::log_err(&err);
+            return message("Login shell exited with nonzero exit code. See logs for details");
+        }
+        Ok(Ok(output)) => output,
+    };
+
+    let Some(parsed_env) = shell_env::parse(&captured).log_err() else {
+        return message("Failed to parse exported environment variables. See logs for the output");
+    };
+    let mut parsed_env = parsed_env
+        .into_iter()
+        .filter_map(|(key, value)| Some((key.to_string(), value?.to_string())))
+        .collect::<HashMap<String, String>>();
+
     // If the user selects `Direct` for direnv, it would set an environment
     // variable that later uses to know that it should not run the hook.
     // We would include in `.envs` call so it is okay to run the hook
     // even if direnv direct mode is enabled.
-    //
-    // In certain shells we need to execute additional_command in order to
-    // trigger the behavior of direnv, etc.
-
-    let command = match shell_name {
-        Some("fish") => format!(
-            "cd '{}'; emit fish_prompt; printf '%s' {MARKER}; /usr/bin/env;",
-            dir.display()
-        ),
-        _ => format!(
-            "cd '{}'; printf '%s' {MARKER}; /usr/bin/env;",
-            dir.display()
-        ),
-    };
-
-    // csh/tcsh only supports `-l` if it's the only flag. So this won't be a login shell.
-    // Users must rely on vars from `~/.tcshrc` or `~/.cshrc` and not `.login` as a result.
-    let args = match shell_name {
-        Some("tcsh") | Some("csh") => vec!["-i".to_string(), "-c".to_string(), command],
-        _ => vec![
-            "-l".to_string(),
-            "-i".to_string(),
-            "-c".to_string(),
-            command,
-        ],
-    };
-
-    let Some(output) = smol::unblock(move || {
-        util::set_pre_exec_to_start_new_session(std::process::Command::new(&shell).args(&args))
-            .output()
-    })
-    .await
-    .log_err() else {
-        return message(
-            "Failed to spawn login shell to source login environment variables. See logs for details",
-        );
-    };
-
-    if !output.status.success() {
-        log::error!("login shell exited with {}", output.status);
-        return message("Login shell exited with nonzero exit code. See logs for details");
-    }
-
-    let stdout = String::from_utf8_lossy(&output.stdout);
-    let Some(env_output_start) = stdout.find(MARKER) else {
-        let stderr = String::from_utf8_lossy(&output.stderr);
-        log::error!(
-            "failed to parse output of `env` command in login shell. stdout: {:?}, stderr: {:?}",
-            stdout,
-            stderr
-        );
-        return message("Failed to parse stdout of env command. See logs for the output");
-    };
-
-    let mut parsed_env = HashMap::default();
-    let env_output = &stdout[env_output_start + MARKER.len()..];
-
-    parse_env_output(env_output, |key, value| {
-        parsed_env.insert(key, value);
-    });
-
     let (direnv_environment, direnv_error) = match load_direnv {
         DirenvSettings::ShellHook => (None, None),
         DirenvSettings::Direct => match load_direnv_environment(&parsed_env, dir).await {
@@ -341,7 +294,6 @@ async fn load_shell_environment(
             ),
         },
     };
-
     for (key, value) in direnv_environment.unwrap_or(HashMap::default()) {
         parsed_env.insert(key, value);
     }

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -175,32 +175,6 @@ impl EnvironmentErrorMessage {
     }
 }
 
-impl From<util::shell_env::Error> for EnvironmentErrorMessage {
-    fn from(err: util::shell_env::Error) -> Self {
-        use util::shell_env::Error;
-        Self::from_str(match err {
-            Error::CannotResolveShellPath(_) => {
-                "Failed to get login environment. SHELL environment variable is not set"
-            }
-            Error::CannotCreateTempfile(_) => {
-                "Failed to create temporary file for environment capture. See logs for details"
-            }
-            Error::SpawnFailed(_) => {
-                "Failed to spawn login shell to source login environment variables. See logs for details"
-            }
-            Error::CannotReadTempfile(_) => {
-                "Failed to read temporary file for environment capture. See logs for details"
-            }
-            Error::ShellExitedWithError(_) => {
-                "Login shell exited with nonzero exit code. See logs for details"
-            }
-            Error::ParseFailed(_) => {
-                "Failed to parse exported environment variables. See logs for the output"
-            }
-        })
-    }
-}
-
 async fn load_directory_shell_environment(
     abs_path: &Path,
     load_direnv: &DirenvSettings,
@@ -275,12 +249,17 @@ async fn load_shell_environment(
     use util::shell_env;
 
     let dir_ = dir.to_owned();
-    let mut envs = match smol::unblock(move || shell_env::capture(Some(dir_)))
-        .await
-        .inspect_err(util::log_err)
-    {
+    let mut envs = match smol::unblock(move || shell_env::capture(Some(dir_))).await {
         Ok(envs) => envs,
-        Err(err) => return (None, Some(err.into())),
+        Err(err) => {
+            util::log_err(&err);
+            return (
+                None,
+                Some(EnvironmentErrorMessage::from_str(
+                    "Failed to load environment variables. See log for details",
+                )),
+            );
+        }
     };
 
     // If the user selects `Direct` for direnv, it would set an environment

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -175,6 +175,32 @@ impl EnvironmentErrorMessage {
     }
 }
 
+impl From<util::shell_env::Error> for EnvironmentErrorMessage {
+    fn from(err: util::shell_env::Error) -> Self {
+        use util::shell_env::Error;
+        Self::from_str(match err {
+            Error::CannotResolveShellPath(_) => {
+                "Failed to get login environment. SHELL environment variable is not set"
+            }
+            Error::CannotCreateTempfile(_) => {
+                "Failed to create temporary file for environment capture. See logs for details"
+            }
+            Error::SpawnFailed(_) => {
+                "Failed to spawn login shell to source login environment variables. See logs for details"
+            }
+            Error::CannotReadTempfile(_) => {
+                "Failed to read temporary file for environment capture. See logs for details"
+            }
+            Error::ShellExitedWithError(_) => {
+                "Login shell exited with nonzero exit code. See logs for details"
+            }
+            Error::ParseFailed(_) => {
+                "Failed to parse exported environment variables. See logs for the output"
+            }
+        })
+    }
+}
+
 async fn load_directory_shell_environment(
     abs_path: &Path,
     load_direnv: &DirenvSettings,
@@ -245,40 +271,17 @@ async fn load_shell_environment(
     Option<HashMap<String, String>>,
     Option<EnvironmentErrorMessage>,
 ) {
-    use crate::direnv::{DirenvError, load_direnv_environment};
+    use crate::direnv::load_direnv_environment;
     use util::shell_env;
 
-    fn message<T>(with: &str) -> (Option<T>, Option<EnvironmentErrorMessage>) {
-        let message = EnvironmentErrorMessage::from_str(with);
-        (None, Some(message))
-    }
-
-    let Some(shell) = std::env::var("SHELL").log_err() else {
-        return message("Failed to get login environment. SHELL environment variable is not set");
-    };
-
     let dir_ = dir.to_owned();
-    let captured = match smol::unblock(move || shell_env::capture(shell, Some(dir_))).await {
-        Err(err) => {
-            util::log_err(&err);
-            return message(
-                "Failed to spawn login shell to source login environment variables. See logs for details",
-            );
-        }
-        Ok(Err(err)) => {
-            util::log_err(&err);
-            return message("Login shell exited with nonzero exit code. See logs for details");
-        }
-        Ok(Ok(output)) => output,
+    let mut envs = match smol::unblock(move || shell_env::capture(Some(dir_)))
+        .await
+        .inspect_err(util::log_err)
+    {
+        Ok(envs) => envs,
+        Err(err) => return (None, Some(err.into())),
     };
-
-    let Some(parsed_env) = shell_env::parse(&captured).log_err() else {
-        return message("Failed to parse exported environment variables. See logs for the output");
-    };
-    let mut parsed_env = parsed_env
-        .into_iter()
-        .filter_map(|(key, value)| Some((key.to_string(), value?.to_string())))
-        .collect::<HashMap<String, String>>();
 
     // If the user selects `Direct` for direnv, it would set an environment
     // variable that later uses to know that it should not run the hook.
@@ -286,19 +289,16 @@ async fn load_shell_environment(
     // even if direnv direct mode is enabled.
     let (direnv_environment, direnv_error) = match load_direnv {
         DirenvSettings::ShellHook => (None, None),
-        DirenvSettings::Direct => match load_direnv_environment(&parsed_env, dir).await {
+        DirenvSettings::Direct => match load_direnv_environment(&envs, dir).await {
             Ok(env) => (Some(env), None),
-            Err(err) => (
-                None,
-                <Option<EnvironmentErrorMessage> as From<DirenvError>>::from(err),
-            ),
+            Err(err) => (None, err.into()),
         },
     };
-    for (key, value) in direnv_environment.unwrap_or(HashMap::default()) {
-        parsed_env.insert(key, value);
+    if let Some(direnv_environment) = direnv_environment {
+        envs.extend(direnv_environment);
     }
 
-    (Some(parsed_env), direnv_error)
+    (Some(envs), direnv_error)
 }
 
 fn get_directory_env_impl(

--- a/crates/util/Cargo.toml
+++ b/crates/util/Cargo.toml
@@ -36,7 +36,6 @@ serde_json_lenient.workspace = true
 smol.workspace = true
 take-until.workspace = true
 tempfile.workspace = true
-thiserror.workspace = true
 unicase.workspace = true
 util_macros = { workspace = true, optional = true }
 walkdir.workspace = true

--- a/crates/util/Cargo.toml
+++ b/crates/util/Cargo.toml
@@ -36,6 +36,7 @@ serde_json_lenient.workspace = true
 smol.workspace = true
 take-until.workspace = true
 tempfile.workspace = true
+thiserror.workspace = true
 unicase.workspace = true
 util_macros = { workspace = true, optional = true }
 walkdir.workspace = true

--- a/crates/util/Cargo.toml
+++ b/crates/util/Cargo.toml
@@ -50,5 +50,6 @@ dunce = "1.0"
 
 [dev-dependencies]
 git2.workspace = true
+indoc.workspace = true
 rand.workspace = true
 util_macros.workspace = true

--- a/crates/util/src/shell_env.rs
+++ b/crates/util/src/shell_env.rs
@@ -1,4 +1,4 @@
-use anyhow::Context as _;
+use anyhow::{Context as _, Result};
 use collections::HashMap;
 use std::borrow::Cow;
 use std::ffi::OsStr;
@@ -7,41 +7,9 @@ use std::path::{Path, PathBuf};
 use std::process::Command;
 use tempfile::NamedTempFile;
 
-pub type Result<T> = std::result::Result<T, Error>;
-
-#[derive(thiserror::Error, Debug)]
-pub enum Error {
-    #[error(
-        "SHELL environment variable is not assigned so we can't source login environment variables"
-    )]
-    CannotResolveShellPath(std::env::VarError),
-
-    #[error(transparent)]
-    CannotCreateTempfile(std::io::Error),
-
-    #[error(transparent)]
-    SpawnFailed(std::io::Error),
-
-    #[error(transparent)]
-    CannotReadTempfile(std::io::Error),
-
-    #[error(
-        "login shell exited with {}. stdout: {:?}, stderr: {:?}",
-        .0.status,
-        String::from_utf8_lossy(&.0.stdout),
-        String::from_utf8_lossy(&.0.stderr),
-    )]
-    ShellExitedWithError(std::process::Output),
-
-    #[error(transparent)]
-    ParseFailed(anyhow::Error),
-}
-
 /// Capture all environment variables from the login shell.
 pub fn capture(change_dir: Option<impl AsRef<Path>>) -> Result<HashMap<String, String>> {
-    let shell_path = std::env::var("SHELL")
-        .map(PathBuf::from)
-        .map_err(Error::CannotResolveShellPath)?;
+    let shell_path = std::env::var("SHELL").map(PathBuf::from)?;
     let shell_name = shell_path.file_name().and_then(OsStr::to_str);
 
     let mut command_string = String::new();
@@ -62,7 +30,7 @@ pub fn capture(change_dir: Option<impl AsRef<Path>>) -> Result<HashMap<String, S
         _ => "",
     });
 
-    let mut env_output_file = NamedTempFile::new().map_err(Error::CannotCreateTempfile)?;
+    let mut env_output_file = NamedTempFile::new()?;
     command_string.push_str(&format!(
         "sh -c 'export -p' > '{}';",
         env_output_file.path().to_string_lossy(),
@@ -81,18 +49,17 @@ pub fn capture(change_dir: Option<impl AsRef<Path>>) -> Result<HashMap<String, S
 
     command.args(["-i", "-c", &command_string]);
 
-    let process_output = super::set_pre_exec_to_start_new_session(&mut command)
-        .output()
-        .map_err(Error::SpawnFailed)?;
-
-    if !process_output.status.success() {
-        return Err(Error::ShellExitedWithError(process_output));
-    }
+    let process_output = super::set_pre_exec_to_start_new_session(&mut command).output()?;
+    anyhow::ensure!(
+        process_output.status.success(),
+        "login shell exited with {}. stdout: {:?}, stderr: {:?}",
+        process_output.status,
+        String::from_utf8_lossy(&process_output.stdout),
+        String::from_utf8_lossy(&process_output.stderr),
+    );
 
     let mut env_output = String::new();
-    env_output_file
-        .read_to_string(&mut env_output)
-        .map_err(Error::CannotReadTempfile)?;
+    env_output_file.read_to_string(&mut env_output)?;
 
     parse(&env_output)
         .filter_map(|entry| match entry {
@@ -123,18 +90,13 @@ fn parse(mut input: &str) -> impl Iterator<Item = Result<(Cow<'_, str>, Option<C
 fn parse_declaration(input: &str) -> Result<((Cow<'_, str>, Option<Cow<'_, str>>), &str)> {
     let rest = input
         .strip_prefix("export ")
-        .context("expected 'export ' prefix")
-        .map_err(Error::ParseFailed)?;
+        .context("expected 'export ' prefix")?;
 
     if let Some((name, rest)) = parse_name_and_terminator(rest, '\n') {
         Ok(((name, None), rest))
     } else {
-        let (name, rest) = parse_name_and_terminator(rest, '=')
-            .context("invalid name")
-            .map_err(Error::ParseFailed)?;
-        let (value, rest) = parse_literal_and_terminator(rest, '\n')
-            .context("invalid value")
-            .map_err(Error::ParseFailed)?;
+        let (name, rest) = parse_name_and_terminator(rest, '=').context("invalid name")?;
+        let (value, rest) = parse_literal_and_terminator(rest, '\n').context("invalid value")?;
         Ok(((name, Some(value)), rest))
     }
 }

--- a/crates/util/src/shell_env.rs
+++ b/crates/util/src/shell_env.rs
@@ -1,13 +1,48 @@
-use anyhow::{Context as _, Result};
-use std::io::{self, Read};
-use std::{borrow::Cow, ffi::OsStr, path::Path, process::Command};
+use anyhow::Context as _;
+use collections::HashMap;
+use std::borrow::Cow;
+use std::ffi::OsStr;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use std::process::Command;
 use tempfile::NamedTempFile;
 
-pub fn capture(
-    shell_path: impl AsRef<Path>,
-    change_dir: Option<impl AsRef<Path>>,
-) -> io::Result<Result<String>> {
-    let shell_name = shell_path.as_ref().file_name().and_then(OsStr::to_str);
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(
+        "SHELL environment variable is not assigned so we can't source login environment variables"
+    )]
+    CannotResolveShellPath(std::env::VarError),
+
+    #[error(transparent)]
+    CannotCreateTempfile(std::io::Error),
+
+    #[error(transparent)]
+    SpawnFailed(std::io::Error),
+
+    #[error(transparent)]
+    CannotReadTempfile(std::io::Error),
+
+    #[error(
+        "login shell exited with {}. stdout: {:?}, stderr: {:?}",
+        .0.status,
+        String::from_utf8_lossy(&.0.stdout),
+        String::from_utf8_lossy(&.0.stderr),
+    )]
+    ShellExitedWithError(std::process::Output),
+
+    #[error(transparent)]
+    ParseFailed(anyhow::Error),
+}
+
+/// Capture all environment variables from the login shell.
+pub fn capture(change_dir: Option<impl AsRef<Path>>) -> Result<HashMap<String, String>> {
+    let shell_path = std::env::var("SHELL")
+        .map(PathBuf::from)
+        .map_err(Error::CannotResolveShellPath)?;
+    let shell_name = shell_path.file_name().and_then(OsStr::to_str);
 
     let mut command_string = String::new();
 
@@ -27,13 +62,13 @@ pub fn capture(
         _ => "",
     });
 
-    let mut env_output_file = NamedTempFile::new()?;
+    let mut env_output_file = NamedTempFile::new().map_err(Error::CannotCreateTempfile)?;
     command_string.push_str(&format!(
         "sh -c 'export -p' > '{}';",
         env_output_file.path().to_string_lossy(),
     ));
 
-    let mut command = Command::new(shell_path.as_ref());
+    let mut command = Command::new(&shell_path);
 
     // For csh/tcsh, the login shell option is set by passing `-` as
     // the 0th argument instead of using `-l`.
@@ -46,45 +81,60 @@ pub fn capture(
 
     command.args(["-i", "-c", &command_string]);
 
-    let process_output = super::set_pre_exec_to_start_new_session(&mut command).output()?;
+    let process_output = super::set_pre_exec_to_start_new_session(&mut command)
+        .output()
+        .map_err(Error::SpawnFailed)?;
 
-    Ok(if process_output.status.success() {
-        let mut env_output = String::new();
-        env_output_file.read_to_string(&mut env_output)?;
-        Ok(env_output)
-    } else {
-        Err(anyhow::anyhow!(
-            "login shell exited with {}. stdout: {:?}, stderr: {:?}",
-            process_output.status,
-            String::from_utf8_lossy(&process_output.stdout),
-            String::from_utf8_lossy(&process_output.stderr),
-        ))
-    })
+    if !process_output.status.success() {
+        return Err(Error::ShellExitedWithError(process_output));
+    }
+
+    let mut env_output = String::new();
+    env_output_file
+        .read_to_string(&mut env_output)
+        .map_err(Error::CannotReadTempfile)?;
+
+    parse(&env_output)
+        .filter_map(|entry| match entry {
+            Ok((name, value)) => Some(Ok((name.into(), value?.into()))),
+            Err(err) => Some(Err(err)),
+        })
+        .collect::<Result<HashMap<String, String>>>()
 }
 
 /// Parse the result of calling `sh -c 'export -p'`.
 ///
 /// https://www.man7.org/linux/man-pages/man1/export.1p.html
-pub fn parse(mut input: &str) -> Result<Vec<(Cow<'_, str>, Option<Cow<'_, str>>)>> {
-    let mut entry;
-    let mut envs = Vec::new();
-    while !input.is_empty() {
-        (entry, input) = parse_declaration(input)?;
-        envs.push(entry);
-    }
-    Ok(envs)
+fn parse(mut input: &str) -> impl Iterator<Item = Result<(Cow<'_, str>, Option<Cow<'_, str>>)>> {
+    std::iter::from_fn(move || {
+        if input.is_empty() {
+            return None;
+        }
+        match parse_declaration(input) {
+            Ok((entry, rest)) => {
+                input = rest;
+                Some(Ok(entry))
+            }
+            Err(err) => Some(Err(err)),
+        }
+    })
 }
 
 fn parse_declaration(input: &str) -> Result<((Cow<'_, str>, Option<Cow<'_, str>>), &str)> {
     let rest = input
         .strip_prefix("export ")
-        .context("expected 'export ' prefix")?;
+        .context("expected 'export ' prefix")
+        .map_err(Error::ParseFailed)?;
 
     if let Some((name, rest)) = parse_name_and_terminator(rest, '\n') {
         Ok(((name, None), rest))
     } else {
-        let (name, rest) = parse_name_and_terminator(rest, '=').context("invalid name")?;
-        let (value, rest) = parse_literal_and_terminator(rest, '\n').context("invalid value")?;
+        let (name, rest) = parse_name_and_terminator(rest, '=')
+            .context("invalid name")
+            .map_err(Error::ParseFailed)?;
+        let (value, rest) = parse_literal_and_terminator(rest, '\n')
+            .context("invalid value")
+            .map_err(Error::ParseFailed)?;
         Ok(((name, Some(value)), rest))
     }
 }
@@ -190,7 +240,7 @@ mod tests {
             Some("b\na\nz"),
             Some("b\\\na\\\nz"),
             Some("baz"),
-            Some(indoc::indoc! { r#"
+            Some(indoc::indoc! {r#"
             \`Hello\`
             \"wo\
             rld\"\n!\\
@@ -204,7 +254,7 @@ mod tests {
             .map(|value| ("foo".into(), value.map(Into::into)))
             .collect::<Vec<_>>();
 
-        let actual = parse(input).unwrap();
+        let actual = parse(input).collect::<Result<Vec<_>>>().unwrap();
         assert_eq!(expected, actual);
     }
 

--- a/crates/util/src/shell_env.rs
+++ b/crates/util/src/shell_env.rs
@@ -1,0 +1,261 @@
+use anyhow::{Context as _, Result};
+use std::io::{self, Read};
+use std::{borrow::Cow, ffi::OsStr, path::Path, process::Command};
+use tempfile::NamedTempFile;
+
+pub fn capture(
+    shell_path: impl AsRef<Path>,
+    change_dir: Option<impl AsRef<Path>>,
+) -> io::Result<Result<String>> {
+    let shell_name = shell_path.as_ref().file_name().and_then(OsStr::to_str);
+
+    let mut command_string = String::new();
+
+    // What we're doing here is to spawn a shell and then `cd` into
+    // the project directory to get the env in there as if the user
+    // `cd`'d into it. We do that because tools like direnv, asdf, ...
+    // hook into `cd` and only set up the env after that.
+    if let Some(dir) = change_dir {
+        let dir_str = dir.as_ref().to_string_lossy();
+        command_string.push_str(&format!("cd '{dir_str}';"));
+    }
+
+    // In certain shells we need to execute additional_command in order to
+    // trigger the behavior of direnv, etc.
+    command_string.push_str(match shell_name {
+        Some("fish") => "emit fish_prompt;",
+        _ => "",
+    });
+
+    let mut env_output_file = NamedTempFile::new()?;
+    command_string.push_str(&format!(
+        "sh -c 'export -p' > '{}';",
+        env_output_file.path().to_string_lossy(),
+    ));
+
+    let mut command = Command::new(shell_path.as_ref());
+
+    // For csh/tcsh, the login shell option is set by passing `-` as
+    // the 0th argument instead of using `-l`.
+    if let Some("tcsh" | "csh") = shell_name {
+        #[cfg(unix)]
+        std::os::unix::process::CommandExt::arg0(&mut command, "-");
+    } else {
+        command.arg("-l");
+    }
+
+    command.args(["-i", "-c", &command_string]);
+
+    let process_output = super::set_pre_exec_to_start_new_session(&mut command).output()?;
+
+    Ok(if process_output.status.success() {
+        let mut env_output = String::new();
+        env_output_file.read_to_string(&mut env_output)?;
+        Ok(env_output)
+    } else {
+        Err(anyhow::anyhow!(
+            "login shell exited with {}. stdout: {:?}, stderr: {:?}",
+            process_output.status,
+            String::from_utf8_lossy(&process_output.stdout),
+            String::from_utf8_lossy(&process_output.stderr),
+        ))
+    })
+}
+
+/// Parse the result of calling `sh -c 'export -p'`.
+///
+/// https://www.man7.org/linux/man-pages/man1/export.1p.html
+pub fn parse(mut input: &str) -> Result<Vec<(Cow<'_, str>, Option<Cow<'_, str>>)>> {
+    let mut entry;
+    let mut envs = Vec::new();
+    while !input.is_empty() {
+        (entry, input) = parse_declaration(input)?;
+        envs.push(entry);
+    }
+    Ok(envs)
+}
+
+fn parse_declaration(input: &str) -> Result<((Cow<'_, str>, Option<Cow<'_, str>>), &str)> {
+    let rest = input
+        .strip_prefix("export ")
+        .context("expected 'export ' prefix")?;
+
+    if let Some((name, rest)) = parse_name_and_terminator(rest, '\n') {
+        Ok(((name, None), rest))
+    } else {
+        let (name, rest) = parse_name_and_terminator(rest, '=').context("invalid name")?;
+        let (value, rest) = parse_literal_and_terminator(rest, '\n').context("invalid value")?;
+        Ok(((name, Some(value)), rest))
+    }
+}
+
+fn parse_name_and_terminator(input: &str, terminator: char) -> Option<(Cow<'_, str>, &str)> {
+    let (name, rest) = parse_literal_and_terminator(input, terminator)?;
+    (!name.is_empty() && !name.contains('=')).then_some((name, rest))
+}
+
+fn parse_literal_and_terminator(input: &str, terminator: char) -> Option<(Cow<'_, str>, &str)> {
+    if let Some((literal, rest)) = parse_literal_single_quoted(input) {
+        let rest = rest.strip_prefix(terminator)?;
+        Some((Cow::Borrowed(literal), rest))
+    } else if let Some((literal, rest)) = parse_literal_double_quoted(input) {
+        let rest = rest.strip_prefix(terminator)?;
+        Some((Cow::Owned(literal), rest))
+    } else {
+        let (literal, rest) = input.split_once(terminator)?;
+        (!literal.contains(|c: char| c.is_ascii_whitespace()))
+            .then_some((Cow::Borrowed(literal), rest))
+    }
+}
+
+/// https://www.gnu.org/software/bash/manual/html_node/Single-Quotes.html
+fn parse_literal_single_quoted(input: &str) -> Option<(&str, &str)> {
+    input.strip_prefix('\'')?.split_once('\'')
+}
+
+/// https://www.gnu.org/software/bash/manual/html_node/Double-Quotes.html
+fn parse_literal_double_quoted(input: &str) -> Option<(String, &str)> {
+    let rest = input.strip_prefix('"')?;
+
+    let mut char_indices = rest.char_indices();
+    let mut escaping = false;
+    let (literal, rest) = loop {
+        let (index, char) = char_indices.next()?;
+        if char == '"' && !escaping {
+            break (&rest[..index], &rest[index + 1..]);
+        } else {
+            escaping = !escaping && char == '\\';
+        }
+    };
+
+    let literal = literal
+        .replace("\\$", "$")
+        .replace("\\`", "`")
+        .replace("\\\"", "\"")
+        .replace("\\\n", "")
+        .replace("\\\\", "\\");
+
+    Some((literal, rest))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_parse() {
+        let input = indoc::indoc! {r#"
+        export foo
+        export 'foo'
+        export "foo"
+        export foo=
+        export 'foo'=
+        export "foo"=
+        export foo=bar
+        export foo='bar'
+        export foo="bar"
+        export foo='b
+        a
+        z'
+        export foo="b
+        a
+        z"
+        export foo='b\
+        a\
+        z'
+        export foo="b\
+        a\
+        z"
+        export foo='\`Hello\`
+        \"wo\
+        rld\"\n!\\
+        !'
+        export foo="\`Hello\`
+        \"wo\
+        rld\"\n!\\
+        !"
+        "#};
+
+        let expected_values = [
+            None,
+            None,
+            None,
+            Some(""),
+            Some(""),
+            Some(""),
+            Some("bar"),
+            Some("bar"),
+            Some("bar"),
+            Some("b\na\nz"),
+            Some("b\na\nz"),
+            Some("b\\\na\\\nz"),
+            Some("baz"),
+            Some(indoc::indoc! { r#"
+            \`Hello\`
+            \"wo\
+            rld\"\n!\\
+            !"#}),
+            Some(indoc::indoc! {r#"
+            `Hello`
+            "world"\n!\!"#}),
+        ];
+        let expected = expected_values
+            .into_iter()
+            .map(|value| ("foo".into(), value.map(Into::into)))
+            .collect::<Vec<_>>();
+
+        let actual = parse(input).unwrap();
+        assert_eq!(expected, actual);
+    }
+
+    #[test]
+    fn test_parse_declaration() {
+        let ((name, value), rest) = parse_declaration("export foo\nrest").unwrap();
+        assert_eq!(name, "foo");
+        assert_eq!(value, None);
+        assert_eq!(rest, "rest");
+
+        let ((name, value), rest) = parse_declaration("export foo=bar\nrest").unwrap();
+        assert_eq!(name, "foo");
+        assert_eq!(value.as_deref(), Some("bar"));
+        assert_eq!(rest, "rest");
+    }
+
+    #[test]
+    fn test_parse_literal_single_quoted() {
+        let input = indoc::indoc! {r#"
+        '\`Hello\`
+        \"wo\
+        rld\"\n!\\
+        !'
+        rest"#};
+
+        let expected = indoc::indoc! {r#"
+        \`Hello\`
+        \"wo\
+        rld\"\n!\\
+        !"#};
+
+        let (actual, rest) = parse_literal_single_quoted(input).unwrap();
+        assert_eq!(expected, actual);
+        assert_eq!(rest, "\nrest");
+    }
+
+    #[test]
+    fn test_parse_literal_double_quoted() {
+        let input = indoc::indoc! {r#"
+        "\`Hello\`
+        \"wo\
+        rld\"\n!\\
+        !"
+        rest"#};
+
+        let expected = indoc::indoc! {r#"
+        `Hello`
+        "world"\n!\!"#};
+
+        let (actual, rest) = parse_literal_double_quoted(input).unwrap();
+        assert_eq!(expected, actual);
+        assert_eq!(rest, "\nrest");
+    }
+}

--- a/crates/util/src/util.rs
+++ b/crates/util/src/util.rs
@@ -28,9 +28,6 @@ use std::{
 };
 use unicase::UniCase;
 
-#[cfg(unix)]
-use anyhow::Context as _;
-
 pub use take_until::*;
 #[cfg(any(test, feature = "test-support"))]
 pub use util_macros::{line_endings, separator, uri};
@@ -313,26 +310,19 @@ fn load_shell_from_passwd() -> Result<()> {
 pub fn load_login_shell_environment() -> Result<()> {
     load_shell_from_passwd().log_err();
 
-    let shell = env::var("SHELL").context(
-        "SHELL environment variable is not assigned so we can't source login environment variables",
-    )?;
-
     // If possible, we want to `cd` in the user's `$HOME` to trigger programs
     // such as direnv, asdf, mise, ... to adjust the PATH. These tools often hook
     // into shell's `cd` command (and hooks) to manipulate env.
     // We do this so that we get the env a user would have when spawning a shell
     // in home directory.
-    let output = shell_env::capture(&shell, env::var_os("HOME"))??;
-
-    for (key, value) in shell_env::parse(&output)? {
-        if let Some(value) = value {
-            unsafe { env::set_var(key.as_ref(), value.as_ref()) };
-        }
+    for (name, value) in shell_env::capture(env::var_os("HOME"))? {
+        unsafe { env::set_var(&name, &value) };
     }
 
     log::info!(
-        "set environment variables from shell:{shell}, path:{}",
-        env::var("PATH").unwrap_or_default(),
+        "set environment variables from shell:{}, path:{}",
+        std::env::var("SHELL").unwrap_or_default(),
+        std::env::var("PATH").unwrap_or_default(),
     );
 
     Ok(())

--- a/crates/util/src/util.rs
+++ b/crates/util/src/util.rs
@@ -5,6 +5,7 @@ pub mod fs;
 pub mod markdown;
 pub mod paths;
 pub mod serde;
+pub mod shell_env;
 pub mod size;
 #[cfg(any(test, feature = "test-support"))]
 pub mod test;
@@ -312,7 +313,6 @@ fn load_shell_from_passwd() -> Result<()> {
 pub fn load_login_shell_environment() -> Result<()> {
     load_shell_from_passwd().log_err();
 
-    let marker = "ZED_LOGIN_SHELL_START";
     let shell = env::var("SHELL").context(
         "SHELL environment variable is not assigned so we can't source login environment variables",
     )?;
@@ -322,35 +322,18 @@ pub fn load_login_shell_environment() -> Result<()> {
     // into shell's `cd` command (and hooks) to manipulate env.
     // We do this so that we get the env a user would have when spawning a shell
     // in home directory.
-    let shell_cmd_prefix = std::env::var_os("HOME")
-        .and_then(|home| home.into_string().ok())
-        .map(|home| format!("cd '{home}';"));
+    let output = shell_env::capture(&shell, env::var_os("HOME"))??;
 
-    let shell_cmd = format!(
-        "{}printf '%s' {marker}; /usr/bin/env;",
-        shell_cmd_prefix.as_deref().unwrap_or("")
-    );
-
-    let output = set_pre_exec_to_start_new_session(
-        std::process::Command::new(&shell).args(["-l", "-i", "-c", &shell_cmd]),
-    )
-    .output()
-    .context("failed to spawn login shell to source login environment variables")?;
-    anyhow::ensure!(output.status.success(), "login shell exited with error");
-
-    let stdout = String::from_utf8_lossy(&output.stdout);
-
-    if let Some(env_output_start) = stdout.find(marker) {
-        let env_output = &stdout[env_output_start + marker.len()..];
-
-        parse_env_output(env_output, |key, value| unsafe { env::set_var(key, value) });
-
-        log::info!(
-            "set environment variables from shell:{}, path:{}",
-            shell,
-            env::var("PATH").unwrap_or_default(),
-        );
+    for (key, value) in shell_env::parse(&output)? {
+        if let Some(value) = value {
+            unsafe { env::set_var(key.as_ref(), value.as_ref()) };
+        }
     }
+
+    log::info!(
+        "set environment variables from shell:{shell}, path:{}",
+        env::var("PATH").unwrap_or_default(),
+    );
 
     Ok(())
 }
@@ -373,32 +356,6 @@ pub fn set_pre_exec_to_start_new_session(
         });
     };
     command
-}
-
-/// Parse the result of calling `usr/bin/env` with no arguments
-pub fn parse_env_output(env: &str, mut f: impl FnMut(String, String)) {
-    let mut current_key: Option<String> = None;
-    let mut current_value: Option<String> = None;
-
-    for line in env.split_terminator('\n') {
-        if let Some(separator_index) = line.find('=') {
-            if !line[..separator_index].is_empty() {
-                if let Some((key, value)) = Option::zip(current_key.take(), current_value.take()) {
-                    f(key, value)
-                }
-                current_key = Some(line[..separator_index].to_string());
-                current_value = Some(line[separator_index + 1..].to_string());
-                continue;
-            };
-        }
-        if let Some(value) = current_value.as_mut() {
-            value.push('\n');
-            value.push_str(line);
-        }
-    }
-    if let Some((key, value)) = Option::zip(current_key.take(), current_value.take()) {
-        f(key, value)
-    }
 }
 
 pub fn merge_json_lenient_value_into(

--- a/crates/util/src/util.rs
+++ b/crates/util/src/util.rs
@@ -315,7 +315,7 @@ pub fn load_login_shell_environment() -> Result<()> {
     // into shell's `cd` command (and hooks) to manipulate env.
     // We do this so that we get the env a user would have when spawning a shell
     // in home directory.
-    for (name, value) in shell_env::capture(env::var_os("HOME"))? {
+    for (name, value) in shell_env::capture(Some(paths::home_dir()))? {
         unsafe { env::set_var(&name, &value) };
     }
 

